### ISSUE TITLE
thor: Panic in AddressSpace destructor code path

### ISSUE THREAD

**Issue by Dennisbonke** [managarm/managarm#224]
There are certain files, mostly shell scripts like `/usr/bin/nroff` and perl scripts like `/usr/bin/autoreconf` that panic frigg on execution.
The following is printed in the e9 log

`automake`
```
posix: Failed to read executable: /bin/automake
frg: Panic! ../../../src/managarm/subprojects/frigg/include/frg/optional.hpp:141: Assertion '_non_n
ll' failed!
```

`nroff`
```
posix: Failed to read executable: /bin/nroff
frg: Panic! ../../../src/managarm/subprojects/frigg/include/frg/optional.hpp:141: Assertion '_non_n
ll' failed!
```

Even a custom script that's only
```
#! /bin/bash

echo "lmaoyeet"
```
fails when executed as `lmaoyeet`. Were one to execute it as `bash /usr/bin/lmaoyeet` it would work fine. Same goes for `nroff` when executed as `bash /usr/bin/nroff` or `automake` when executed as `perl /usr/bin/automake` or whatever the scripts shebang may be

**Comment by Dennisbonke** [comment]
a `gdb` backtrace gives the following
```
(gdb) bt
#0  thor::panic () at ../../../src/managarm/kernel/thor/generic/debug.cpp:8
#1  thor::PanicSink::operator() (this=<optimized out>, msg=<optimized out>)
    at ../../../src/managarm/kernel/thor/generic/debug.cpp:31
#2  0xffffffff80012052 in frg::stack_buffer_logger<thor::PanicSink, 128ul>::_emit (
    this=0xffffffff8012c1c4 <thor::logMutex>, 
    message=0xffffe00000b3fdb8 "frg: Panic! ../../../src/managarm/subprojects/frigg/include/frg/optional.hpp:141: Assertion '_non_null' failed!")
    at ../../../src/managarm/subprojects/frigg/include/frg/logging.hpp:80
#3  frg::stack_buffer_logger<thor::PanicSink, 128ul>::item::operator<< (this=0xffffe00000b3fdb0)
    at ../../../src/managarm/subprojects/frigg/include/frg/logging.hpp:37
#4  frg_panic (cstring=<optimized out>) at ../../../src/managarm/kernel/thor/generic/main.cpp:45
#5  0xffffffff80008948 in frg::optional<bool>::operator* (this=<optimized out>)
    at ../../../src/managarm/subprojects/frigg/include/frg/optional.hpp:141
#6  async::sender_awaiter<async::oneshot_event::wait_sender, bool>::await_resume (this=<optimized out>)
    at ../../../src/managarm/subprojects/libasync/include/async/basic.hpp:73
#7  thor::VirtualSpace::retire()::$_3::operator()(smarter::shared_ptr<thor::VirtualSpace, void>) const (
    this=<optimized out>, self=...) at ../../../src/managarm/kernel/thor/generic/address-space.cpp:346
#8  0xffffffff8004449c in smarter::counter::decrement (this=0xffffe00005180a30)
    at ../../../src/managarm/subprojects/libsmarter/include/smarter.hpp:85
#9  smarter::shared_ptr<thor::AddressSpace, BindableHandle>::~shared_ptr (this=0xffffe00000b3fed0)
    at ../../../src/managarm/subprojects/libsmarter/include/smarter.hpp:279
#10 thor::AddressSpaceDescriptor::~AddressSpaceDescriptor (this=0xffffe00000b3fed0)
    at ../../../src/managarm/kernel/thor/generic/thor-internal/descriptor.hpp:59
#11 frg::variant<thor::UniverseDescriptor, thor::QueueDescriptor, thor::MemoryViewDescriptor, thor::MemorySliceDescriptor, thor::AddressSpaceDescriptor, thor::VirtualizedSpaceDescriptor, thor::VirtualizedCpuDescriptor, thor::MemoryViewLockDescriptor, thor::ThreadDescriptor, thor::LaneDescriptor, thor::IrqDescriptor, thor::OneshotEventDescriptor, thor::BitsetEventDescriptor, thor::IoDescriptor, thor::KernletObjectDescriptor, thor::BoundKernletDescriptor>::destruct_<4ul> (this=<optimized out>)
    at ../../../src/managarm/subprojects/frigg/include/frg/variant.hpp:199
#12 frg::variant<thor::UniverseDescriptor, thor::QueueDescriptor, thor::MemoryViewDescriptor, thor::MemorySliceDescriptor, thor::AddressSpaceDescriptor, thor::VirtualizedSpaceDescriptor, thor::VirtualizedCpuDescriptor, thor::MemoryViewLockDescriptor, thor::ThreadDescriptor, thor::LaneDescriptor, thor::IrqDescriptor, thor::OneshotEventDescriptor, thor::BitsetEventDescriptor, thor::IoDescriptor, thor::KernletObjectDescriptor, thor::BoundKernletDescriptor>::destruct_<3ul> (this=0xffffe00000b3fec8)
    at ../../../src/managarm/subprojects/frigg/include/frg/variant.hpp:202
#13 0xffffffff800443f4 in frg::variant<thor::UniverseDescriptor, thor::QueueDescriptor, thor::MemoryViewDescriptor, thor::MemorySliceDescriptor, thor::AddressSpaceDescriptor, thor::VirtualizedSpaceDescriptor, thor::VirtualizedCpuDescriptor, thor::MemoryViewLockDescriptor, thor::ThreadDescriptor, thor::LaneDescriptor, thor::IrqDescriptor, thor::OneshotEventDescriptor, thor::BitsetEventDescriptor, thor::IoDescriptor, thor::KernletObjectDescriptor, thor::BoundKernletDescriptor>::destruct_<2ul> (this=<optimized out>)
    at ../../../src/managarm/subprojects/frigg/include/frg/variant.hpp:202
#14 frg::variant<thor::UniverseDescriptor, thor::QueueDescriptor, thor::MemoryViewDescriptor, thor::MemorySliceDescriptor, thor::AddressSpaceDescriptor, thor::VirtualizedSpaceDescriptor, thor::VirtualizedCpuDescriptor, thor::MemoryViewLockDescriptor, thor::ThreadDescriptor, thor::LaneDescriptor, thor::IrqDescriptor, thor::OneshotEventDescriptor, thor::BitsetEventDescriptor, thor::IoDescriptor, thor::KernletObjectDescriptor, thor::BoundKernletDescriptor>::destruct_<1ul> (this=<optimized out>)
    at ../../../src/managarm/subprojects/frigg/include/frg/variant.hpp:202
#15 0xffffffff80056b30 in frg::variant<thor::UniverseDescriptor, thor::QueueDescriptor, thor::MemoryViewDescriptor, thor::MemorySliceDescriptor, thor::AddressSpaceDescriptor, thor::VirtualizedSpaceDescriptor, thor::VirtualizedCpuDescriptor, thor::MemoryViewLockDescriptor, thor::ThreadDescriptor, thor::LaneDescriptor, thor::IrqDescriptor, thor::OneshotEventDescriptor, thor::BitsetEventDescriptor, thor::IoDescriptor, t--Type <RET> for more, q to quit, c to continue without paging--
hor::KernletObjectDescriptor, thor::BoundKernletDescriptor>::destruct_<0ul> (this=<optimized out>)
    at ../../../src/managarm/subprojects/frigg/include/frg/variant.hpp:202
#16 frg::variant<thor::UniverseDescriptor, thor::QueueDescriptor, thor::MemoryViewDescriptor, thor::MemorySliceDescriptor, thor::AddressSpaceDescriptor, thor::VirtualizedSpaceDescriptor, thor::VirtualizedCpuDescriptor, thor::MemoryViewLockDescriptor, thor::ThreadDescriptor, thor::LaneDescriptor, thor::IrqDescriptor, thor::OneshotEventDescriptor, thor::BitsetEventDescriptor, thor::IoDescriptor, thor::KernletObjectDescriptor, thor::BoundKernletDescriptor>::~variant (this=0xffffe00000b3fec8)
    at ../../../src/managarm/subprojects/frigg/include/frg/variant.hpp:81
#17 frg::optional<frg::variant<thor::UniverseDescriptor, thor::QueueDescriptor, thor::MemoryViewDescriptor, thor::MemorySliceDescriptor, thor::AddressSpaceDescriptor, thor::VirtualizedSpaceDescriptor, thor::VirtualizedCpuDescriptor, thor::MemoryViewLockDescriptor, thor::ThreadDescriptor, thor::LaneDescriptor, thor::IrqDescriptor, thor::OneshotEventDescriptor, thor::BitsetEventDescriptor, thor::IoDescriptor, thor::KernletObjectDescriptor, thor::BoundKernletDescriptor> >::_reset (this=0xffffe00000b3fec8)
    at ../../../src/managarm/subprojects/frigg/include/frg/optional.hpp:170
#18 frg::optional<frg::variant<thor::UniverseDescriptor, thor::QueueDescriptor, thor::MemoryViewDescriptor, thor::MemorySliceDescriptor, thor::AddressSpaceDescriptor, thor::VirtualizedSpaceDescriptor, thor::VirtualizedCpuDescriptor, thor::MemoryViewLockDescriptor, thor::ThreadDescriptor, thor::LaneDescriptor, thor::IrqDescriptor, thor::OneshotEventDescriptor, thor::BitsetEventDescriptor, thor::IoDescriptor, thor::KernletObjectDescriptor, thor::BoundKernletDescriptor> >::~optional (this=0xffffe00000b3fec8)
    at ../../../src/managarm/subprojects/frigg/include/frg/optional.hpp:57
#19 helCloseDescriptor (universeHandle=<optimized out>, handle=<optimized out>)
    at ../../../src/managarm/kernel/thor/generic/hel.cpp:243
#20 0xffffffff800164c6 in thor::handleSyscall (image=...)
    at ../../../src/managarm/kernel/thor/generic/main.cpp:605
#21 0xffffffff800dadd3 in onPlatformSyscall (image=...)
    at ../../../src/managarm/kernel/thor/arch/x86/ints.cpp:438
#22 0xffffffff80001fc1 in syscallStub () at ../../../src/managarm/kernel/thor/arch/x86/stubs.S:487
#23 0x0000000000000015 in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
```

**Comment by avdgrinten** [comment]
This should not be closed by #225: while that PR fixes a bug in POSIX, the kernel panic needs to be dealt with separately.
